### PR TITLE
Add support for macros with new `swift_compiler_plugin` rule

### PR DESCRIPTION
**Pull request by keith**
This adds support for macros with a new swift_compiler_plugin rule. Example usage:

```bzl
swift_compiler_plugin(
    name = "Macros",
    srcs = glob(["Macros/*.swift"]),
    deps = [
        "@SwiftSyntax",
        "@SwiftSyntax//:SwiftCompilerPlugin",
        "@SwiftSyntax//:SwiftSyntaxMacros",
    ],
)

swift_test(
    name = "MacrosTests",
    srcs = glob(["MacrosTests/*.swift"]),
    deps = [
        ":Macros",
        "@SwiftSyntax//:SwiftSyntaxMacrosTestSupport",
    ],
)

swift_library(
    name = "MacroLibrary",
    srcs = glob(["MacroLibrary/*.swift"]),
    plugins = [":Macros"],
)

swift_library(
    name = "MacroConsumer",
    srcs = glob(["Sources/*.swift"]),
    deps = [":MacroLibrary"],
)
```

### CODE REVIEW

**Comment by keith** [comment]
might need https://github.com/bazelbuild/rules_swift/commit/69349eeba42ed7b7365759ba208067b4aa23854d

**Comment by brentleyjones** [comment]
I'll update my local usage of this tomorrow to test it out.

File: doc/BUILD.bazel [hunk 1]
```diff
@@ -18,7 +18,7 @@ _DOC_SRCS = {
     "rules": [
         "swift_binary",
         "swift_c_module",
-        "swift_feature_allowlist",
```
Comment (brentleyjones) [inline doc/BUILD.bazel:21]: Is the removal here expected?

**Comment by brentleyjones** [comment]
Local testing worked.

File: doc/BUILD.bazel (hunk 1 above)
Comment (keith) [reply doc/BUILD.bazel:21]: added back

**Comment by keith** [comment]
added a feature check inside `swift_compiler_plugin` to fail early in that case. that should be enough since it's the only thing that creates the provider required for use in `plugins = [ ... ]`

File: swift/internal/compiling.bzl [hunk 2]
```diff
@@ -2524,6 +2594,7 @@ def compile(
         ast_files = compile_outputs.ast_files,
         indexstore = compile_outputs.indexstore_directory,
         symbol_graph = compile_outputs.symbol_graph_directory,
+        macro_expansion_directory = compile_outputs.macro_expansion_directory,
```
Comment (brentleyjones) [inline swift/internal/compiling.bzl:2597]: Should be moved up to stay alphabetical?

File: swift/internal/swift_binary_test.bzl [hunk 3]
```diff
@@ -210,7 +107,9 @@ def _swift_linking_rule_impl(
             is_test = ctx.attr.testonly,
             module_name = module_name,
             package_name = ctx.attr.package_name,
+            plugins = get_providers(ctx.attr.plugins, SwiftCompilerPluginInfo),
             srcs = srcs,
+            extra_swift_infos = extra_swift_infos,
```
Comment (brentleyjones) [inline swift/internal/swift_binary_test.bzl:112]: Can this be moved up to stay alphabetical?

File: swift/internal/compiling.bzl (hunk 2 above)
Comment (keith) [reply swift/internal/compiling.bzl:2596]: done

File: swift/internal/swift_binary_test.bzl (hunk 3 above)
Comment (keith) [reply swift/internal/swift_binary_test.bzl]: done

**Review by brentleyjones — APPROVED** [review]

**Pull request merged**